### ISSUE TITLE
GroupedBatchSampler in references has wrong __len__

### ISSUE THREAD

**Issue by davidgill97** [pytorch/vision#9127]
### 🐛 Describe the bug

[GroupedBatchSampler](https://github.com/pytorch/vision/blob/6aee5eddc8f9b397b45d076bd336c030518b7626/references/detection/group_by_aspect_ratio.py#L22C7-L22C26) has wrong `__len__`. Current `__len__` is implemented as 
```python
    def __len__(self):
        return len(self.sampler) // self.batch_size
```
This causes the last batch to be dropped when the number of samples is not divisible by batch size. It should be corrected as 
```python
    def __len__(self):
        return math.ceil(len(self.sampler) / self.batch_size)
```
to fix the problem.

### Versions

Collecting environment information...
PyTorch version: 2.7.1+cu128
Is debug build: False
CUDA used to build PyTorch: 12.8
ROCM used to build PyTorch: N/A

OS: Ubuntu 20.04.6 LTS (x86_64)
GCC version: (Ubuntu 9.4.0-1ubuntu1~20.04.2) 9.4.0
Clang version: Could not collect
CMake version: Could not collect
Libc version: glibc-2.31

Python version: 3.10.18 (main, Jun  4 2025, 08:56:00) [GCC 9.4.0] (64-bit runtime)
Python platform: Linux-6.11.0-25-generic-x86_64-with-glibc2.31
Is CUDA available: True
CUDA runtime version: 12.8.93
CUDA_MODULE_LOADING set to: LAZY
GPU models and configuration: 
GPU 0: NVIDIA GeForce RTX 5070
GPU 1: NVIDIA GeForce RTX 5070

Nvidia driver version: 570.144
cuDNN version: Probably one of the following:
/usr/lib/x86_64-linux-gnu/libcudnn.so.9.8.0
/usr/lib/x86_64-linux-gnu/libcudnn_adv.so.9.8.0
/usr/lib/x86_64-linux-gnu/libcudnn_cnn.so.9.8.0
/usr/lib/x86_64-linux-gnu/libcudnn_engines_precompiled.so.9.8.0
/usr/lib/x86_64-linux-gnu/libcudnn_engines_runtime_compiled.so.9.8.0
/usr/lib/x86_64-linux-gnu/libcudnn_graph.so.9.8.0
/usr/lib/x86_64-linux-gnu/libcudnn_heuristic.so.9.8.0
/usr/lib/x86_64-linux-gnu/libcudnn_ops.so.9.8.0
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

CPU:
Architecture:                         x86_64
CPU op-mode(s):                       32-bit, 64-bit
Byte Order:                           Little Endian
Address sizes:                        46 bits physical, 48 bits virtual
CPU(s):                               24
On-line CPU(s) list:                  0-23
Thread(s) per core:                   1
Core(s) per socket:                   24
Socket(s):                            1
NUMA node(s):                         1
Vendor ID:                            GenuineIntel
CPU family:                           6
Model:                                198
Model name:                           Intel(R) Core(TM) Ultra 9 285K
Stepping:                             2
CPU MHz:                              2162.622
CPU max MHz:                          5200.0000
CPU min MHz:                          800.0000
BogoMIPS:                             7372.80
Virtualization:                       VT-x
L1d cache:                            1.1 MiB
L1i cache:                            1.5 MiB
L2 cache:                             72 MiB
L3 cache:                             36 MiB
NUMA node0 CPU(s):                    0-23
Vulnerability Gather data sampling:   Not affected
Vulnerability Itlb multihit:          Not affected
Vulnerability L1tf:                   Not affected
Vulnerability Mds:                    Not affected
Vulnerability Meltdown:               Not affected
Vulnerability Mmio stale data:        Not affected
Vulnerability Reg file data sampling: Not affected
Vulnerability Retbleed:               Not affected
Vulnerability Spec rstack overflow:   Not affected
Vulnerability Spec store bypass:      Mitigation; Speculative Store Bypass disabled via prctl
Vulnerability Spectre v1:             Mitigation; usercopy/swapgs barriers and __user pointer sanitization
Vulnerability Spectre v2:             Mitigation; Enhanced / Automatic IBRS; IBPB conditional; RSB filling; PBRSB-eIBRS Not affected; BHI Not affected
Vulnerability Srbds:                  Not affected
Vulnerability Tsx async abort:        Not affected
Flags:                                fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush dts acpi mmx fxsr sse sse2 ss ht tm pbe syscall nx pdpe1gb rdtscp lm constant_tsc art arch_perfmon pebs bts rep_good nopl xtopology nonstop_tsc cpuid aperfmperf tsc_known_freq pni pclmulqdq dtes64 monitor ds_cpl vmx smx est tm2 ssse3 sdbg fma cx16 xtpr pdcm pcid sse4_1 sse4_2 x2apic movbe popcnt tsc_deadline_timer aes xsave avx f16c rdrand lahf_lm abm 3dnowprefetch cpuid_fault ssbd ibrs ibpb stibp ibrs_enhanced tpr_shadow flexpriority ept vpid ept_ad fsgsbase tsc_adjust bmi1 avx2 smep bmi2 erms invpcid rdt_a rdseed adx smap clflushopt clwb intel_pt sha_ni xsaveopt xsavec xgetbv1 xsaves split_lock_detect user_shstk avx_vnni lam wbnoinvd dtherm ida arat pln pts hwp hwp_notify hwp_act_window hwp_epp hwp_pkg_req hfi vnmi umip pku ospke waitpkg gfni vaes vpclmulqdq rdpid bus_lock_detect movdiri movdir64b fsrm md_clear serialize arch_lbr ibt flush_l1d arch_capabilities

Versions of relevant libraries:
[pip3] numpy==2.2.6
[pip3] nvidia-cublas-cu12==12.8.3.14
[pip3] nvidia-cuda-cupti-cu12==12.8.57
[pip3] nvidia-cuda-nvrtc-cu12==12.8.61
[pip3] nvidia-cuda-runtime-cu12==12.8.57
[pip3] nvidia-cudnn-cu12==9.7.1.26
[pip3] nvidia-cufft-cu12==11.3.3.41
[pip3] nvidia-curand-cu12==10.3.9.55
[pip3] nvidia-cusolver-cu12==11.7.2.55
[pip3] nvidia-cusparse-cu12==12.5.7.53
[pip3] nvidia-cusparselt-cu12==0.6.3
[pip3] nvidia-nccl-cu12==2.26.2
[pip3] nvidia-nvjitlink-cu12==12.8.61
[pip3] nvidia-nvtx-cu12==12.8.55
[pip3] onnx==1.18.0
[pip3] onnxruntime-gpu==1.21.0
[pip3] onnxsim==0.4.36
[pip3] pytorch-lightning==2.5.2
[pip3] torch==2.7.1+cu128
[pip3] torchmetrics==1.7.3
[pip3] torchvision==0.22.1+cu128
[pip3] triton==3.3.1
[conda] Could not collect

**Comment by scotts** [comment]
@davidgill97, I think you're correct, as there as the remaining logic will always produce batches of `batch_size`, repeating elements if necessary: https://github.com/pytorch/vision/blob/6aee5eddc8f9b397b45d076bd336c030518b7626/references/detection/group_by_aspect_ratio.py#L65-L77
So if `len(self.sampler) == 101` and `self.batch_size == 33`, then we want to produce 34 total batches. The first 33 batches will contain unique indices, but the 34th batch will contain the last index repeated `self.bath_size` times. The code as currently implemented, however, will return 33 as the number of batches. That is, we're currently getting the floor, but we should get the ceiling.

However, the object detection training scripts are a part of TorchVision we are not actively improving; see #9036. We do welcome a tested PR if you would fix this behavior.

**Comment by NicolasHug** [comment]
I suspect this was intentional, it's usually important for all the batch sizes to be consistently equal. Some ops, like BatchNorm, are sensitive to the batch size being constant. I'll close this issue @davidgill97 please follow-up if you think this is a genuine bug.
